### PR TITLE
JDK-8260388: Listing (sub)packages at package level of API documentation

### PR DESCRIPTION
**Pull request by hns**
This adds a "Related Packages" table to package summary pages with a list of neighboring tables. The rules for including packages is as follows:

 1. The super package of the current package is included if it exists.
 2. Direct subpackages of the current package are included, but only if their number does not exceed 20, in which case no subpackages are included.
 3. If steps 1 and 2 did not produce more than 5 packages, sibling packages of the current package are included, but only if their number is not greater than 5.

In steps 2. and 3. above, either all or no candidate packages are included. In other words, if the threshold is exceeded none of the candidate packages are included in the list. The cut-off values are arbitrary, but have shown to produce reasonably good results with various codebases.

A few example of API documentation generated with this feature:

http://cr.openjdk.java.net/~hannesw/8260388/api.01/
http://cr.openjdk.java.net/~hannesw/8260388/javafx.01/
http://cr.openjdk.java.net/~hannesw/8260388/tomcat.02/
http://cr.openjdk.java.net/~hannesw/8260388/servlet.01/

This is a first step to introduce the new feature. There may be future enhancements to fine-tune the feature or make it more customizable.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260388](https://bugs.openjdk.java.net/browse/JDK-8260388): Listing (sub)packages at package level of API documentation

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to 7e9a6e07058f19bbdd63d3fb39674261697c610e

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/2917/head:pull/2917`
`$ git checkout pull/2917`

To update a local copy of the PR:
`$ git checkout pull/2917`
`$ git pull https://git.openjdk.java.net/jdk pull/2917/head`

### CODE REVIEW

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/builders/PackageSummaryBuilder.java [hunk 1]
```diff
@@ -57,6 +62,10 @@
      */
     private final PackageSummaryWriter packageWriter;
 
+    // Maximum number of subpackages and sibling packages to list in related packages table
+    private final static int MAX_SUBPACKAGES = 20;
+    private final static int MAX_SIBLING_PACKAGES = 5;
+
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/builders/PackageSummaryBuilder.java:68]: For future work, move these to a config file.

Short-term "workaround" suggestions still for a separate RFR: annotate these and similar declarations with an internal source-retention annotation of `@Configuration`

**Review by jonathan-gibbons — APPROVED** [review]
All OK; some suggestions for future work/cleanup/improvement

**Comment by hns** [comment]
I have added a change to this PR to add a "Module" column to the related packages table if any of the packages is in a different module. Example output for the JDK libraries is available here (`java.lang`, `java.util` and `com.sun.tools.*` packages have a module column):

http://cr.openjdk.java.net/~hannesw/8260388/api.03/

Implementation note: I added a HtmlStyle.colExtra CSS style because existing column classes either had unwanted rules (such as displaying links in bold font) or didn't fit for other reasons.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/PackageWriterImpl.java [hunk 2]
```diff
@@ -235,6 +245,49 @@ public void addClassesSummary(SortedSet<TypeElement> classes, String label,
         }
     }
 
+    public void addPackageSummary(List<PackageElement> packages, Content label,
+                                  TableHeader tableHeader, Content summaryContentTree,
+                                  boolean showModules) {
+        if(!packages.isEmpty()) {
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/PackageWriterImpl.java:251]: whitespace after `if`

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/PackageWriterImpl.java [hunk 3]
```diff
@@ -235,6 +245,49 @@ public void addClassesSummary(SortedSet<TypeElement> classes, String label,
...
+                    description.add(getDeprecatedPhrase(pkg));
+                    List<? extends DeprecatedTree> tags = utils.getDeprecatedTrees(pkg);
+                    if (!tags.isEmpty()) {
+                        addSummaryDeprecatedComment(pkg, tags.get(0), description);
+                    }
+                } else {
+                    addSummaryComment(pkg, description);
+                }
+                if (showModules) {
+                    table.addRow(packageLink, moduleLink, description);
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/PackageWriterImpl.java:282]: This looks like the companion row to the header entry, to put the module column first.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlStyle.java [hunk 4]
```diff
@@ -441,6 +441,12 @@
      */
     colSummaryItemName,
 
+    /**
+     * The class of the cells in a table column used to display additional
+     * information without any particular style.
+     */
+    colExtra,
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlStyle.java:448]: The `extra` in the same seems irrelevant. How about `colPlain`, to indicate it is a column with no additional style.

**Review by jonathan-gibbons — APPROVED** [review]
Definitely better with the new column for the linked module name.

This is OK as is, but I think it would be better to have the module column first, before the package column.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlStyle.java (hunk 4 above)
Comment (hns) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlStyle.java:448]: I was considering `colPlain`, but dismissed it because I thought we or somebody else might actually add styles to it at some point. But I'll take the risk of future cognitive dissonance and go with `colPlain`.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/PackageWriterImpl.java [hunk 5]
```diff
@@ -162,6 +162,16 @@ public Content getSummariesList() {
         return new HtmlTree(TagName.UL).setStyle(HtmlStyle.summaryList);
     }
 
+    @Override
+    public void addRelatedPackagesSummary(List<PackageElement> relatedPackages, Content summaryContentTree) {
+        boolean showModules = configuration.showModules && hasRelatedPackagesInOtherModules(relatedPackages);
+        TableHeader tableHeader= showModules
+                ? new TableHeader(contents.packageLabel, contents.moduleLabel, contents.descriptionLabel)
```
[the remark this answers is not in the document]
Comment (hns) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/PackageWriterImpl.java:169]: I tried it out and was very surprised to find I agree with you. I'm changing the order in a new commit.
